src/MongoDB.Driver.Core/Core/Authentication/MongoAWSAuthenticator.cs

```diff
@@ -0,0 +1,531 @@
+        {
+            string sessionToken = null;
+            if (properties != null)
+            {
+                foreach (var pair in properties)
+                {
+                    switch (pair.Key.ToUpperInvariant())
+                    {
+                        case "AWS_SESSION_TOKEN":
+                            sessionToken = (string)pair.Value;
+                            break;
+                        default:
+                            throw new ArgumentException($"Unknown AWS property '{pair.Key}'.", nameof(properties));
+                    }
+                }
+            }
+
+            return sessionToken;
+        }
+
+        private static SecureString ToSecureString(string str)
+        {
+            if (str == null)
+            {
+                return null;
+            }
+
+            var secureString = new SecureString();
+            foreach (var c in str)
+            {
+                secureString.AppendChar(c);
+            }
+            secureString.MakeReadOnly();
+
+            return secureString;
+        }
+
+        private static void ValidateAwsCredentials(AwsCredentials awsCredentials)
+        {
+            if (awsCredentials.Username == null && (awsCredentials.Password != null || awsCredentials.SessionToken != null))
+            {
+                throw new ArgumentException("A MONGODB-AWS must have access key id.");
+            }
+            if (awsCredentials.Username != null && awsCredentials.Password == null)
+            {
+                throw new ArgumentException("A MONGODB-AWS must have secret access key.");
+            }
+        }
+
+        private static BsonDocument ToDocument(byte[] bytes)
+        {
+            MemoryStream stream = new MemoryStream(bytes);
+            using (var jsonReader = new BsonBinaryReader(stream))
+            {
+                var context = BsonDeserializationContext.CreateRoot(jsonReader);
+                return BsonDocumentSerializer.Instance.Deserialize(context);
+            }
+        }
+
+        private static byte[] ToBytes(BsonDocument doc)
+        {
+#pragma warning disable 618
+            if (BsonDefaults.GuidRepresentationMode == GuidRepresentationMode.V2)
+            {
+                return doc.ToBson(writerSettings: new BsonBinaryWriterSettings
+                {
+                    GuidRepresentation = GuidRepresentation.Unspecified
+                });
+            }
+#pragma warning restore 618
+
+            return doc.ToBson();
+        }
+
+        // nested classes
+        private class AwsCredentials
+        {
+            public string Username;
+            public SecureString Password;
+            public string SessionToken;
+        }
+
+        private class AwsHttpClientHelper
+        {
+            // private static
+            private static readonly Uri _ec2BaseUri = new Uri("http://169.254.169.254");
```

Comment:
nit: rename to `__ecs2BaseUri` 

We use two leading underscores to distinguish static fields from instance fields.